## Practice Case PyMongo
**AUTHOR : HANIA MAGHFIRA**

Connect cluster pymongo pada link berikut (pilih salah satu yang bisa) :
- **mongodb+srv://userstudent:admin1234@cluster0-nnbxe.gcp.mongodb.net/test?retryWrites=true&w=majority** (used)
- mongodb+srv://mongodb-intro:admin1234@cluster0-nnbxe.gcp.mongodb.net/test?retryWrites=true&w=majority
- mongodb+srv://adminuser12345:adminuser12345@cluster0-phmf4.gcp.mongodb.net/test?retryWrites=true&w=majority
- mongodb+srv://admin12345:admin123@cluster0-snbhz.gcp.mongodb.net/test?retryWrites=true&w=majority
- **mongodb+srv://user1234:user1234@cluster0-1dcm4.gcp.mongodb.net/test?retryWrites=true&w=majority** (used)
- mongodb+srv://admin1234:12345@cluster0-miqju.gcp.mongodb.net/test?retryWrites=true&w=majority

#### Question
1. Membuat collection baru dengan nama clean_movies yang sama persis dengan movies collection pada database sample_mflix dengan memakai collection movies_intial (projecting_queries pymongo)
2. Validasi collection yang telah dibuat dengan parameter sbb :
    1. Semua document pada clean_movies dan movie sama
    2. Banyak document pada clean_movies dan movie sama
    3. Semua fields pada clean_movies ada pada movie
    4. Semua value pada clean_movies sama dengan semua value pada movies dengan urutan yang sama

---

# Install Dependencies

In [35]:
# install pymongo
!pip install pymongo[srv]

# install pprint
!pip install pprint

You are using pip version 9.0.1, however version 19.3.1 is available.
You should consider upgrading via the 'python -m pip install --upgrade pip' command.


You are using pip version 9.0.1, however version 19.3.1 is available.
You should consider upgrading via the 'python -m pip install --upgrade pip' command.


# Import Library

In [17]:
# import library
import pymongo
from pymongo import MongoClient
import pprint
from IPython.display import clear_output

# Connect to MongoDB
Firstly, we need to connect to MongoDB which have movies collection to match the values and connect to another MongoDB to store new collection.<br>

**client** : connection to MongoDB which have movies collection.<br>
**client_movies** : connection to store new collection.

In [18]:
# connect to MongoDB
client = MongoClient("mongodb+srv://userstudent:admin1234@cluster0-nnbxe.gcp.mongodb.net/test?retryWrites=true&w=majority") #main cluster where movies is located
client_movies = MongoClient("mongodb+srv://user1234:user1234@cluster0-1dcm4.gcp.mongodb.net/test?retryWrites=true&w=majority") #cluster to store new collection

# Connect to sample_mflix database to get movies collection

In [41]:
# show db sample_mflix
db = client.sample_mflix #database
col_movies = db.movies #movies collection
col_movies_initial = db.movies_initial #movies_initial collection

In [45]:
# show movies collection
pprint.pprint(col_movies.find_one())

{'_id': ObjectId('573a1390f29313caabcd5a93'),
 'awards': {'nominations': 0, 'text': '1 win.', 'wins': 1},
 'cast': ['Howard C. Hickman', 'Enid Markey', 'Lola May', 'Kate Bruce'],
 'countries': ['USA'],
 'directors': ['Reginald Barker',
               'Thomas H. Ince',
               'Raymond B. West',
               'Walter Edwards',
               'David Hartford',
               'Jay Hunt',
               'J. Parker Read Jr.'],
 'fullplot': 'Allegorical film about peace. A king starts a war, many of the '
             'women are against it, people are pressed into service. A count '
             'has constructed a submarine and gets the order to sink an ocean '
             'liner, that is also carrying - supposedly - ammunition for the '
             'enemy. The count refuses to fire the torpedos, and sinks the '
             'submarine. He survives, but in a limbo between death and life '
             'where he meets Jesus, who takes him over to preach peace. '
             'Natura

In [46]:
# show movies_initial collection
pprint.pprint(col_movies_initial.find_one())

{'_id': ObjectId('5daf3e85ab0f2781ea55d595'),
 'awards': '',
 'cast': 'Carmencita',
 'country': 'USA',
 'director': 'William K.L. Dickson',
 'fullplot': 'Performing on what looks like a small wooden stage, wearing a '
             'dress with a hoop skirt and white high-heeled pumps, Carmencita '
             'does a dance with kicks and twirls, a smile always on her face.',
 'genre': 'Documentary, Short',
 'imdbID': 1,
 'imdbRating': 5.9,
 'imdbVotes': 1032,
 'language': '',
 'lastupdated': '2015-08-26 00:03:45.040000000',
 'metacritic': '',
 'plot': 'Performing on what looks like a small wooden stage, wearing a dress '
         'with a hoop skirt and white high-heeled pumps, Carmencita does a '
         'dance with kicks and twirls, a smile always on her face.',
 'poster': 'http://ia.media-imdb.com/images/M/MV5BMjAzNDEwMzk3OV5BMl5BanBnXkFtZTcwOTk4OTM5Ng@@._V1_SX300.jpg',
 'rating': 'NOT RATED',
 'released': '',
 'runtime': '1 min',
 'title': 'Carmencita',
 'type': 'movie',
 'writer':

In [36]:
# matching imdb in movies collection
# so that the number of document in new collection is equal to number of document in movies collection
imdb = list(col_movies.find({}, {'_id':0, 'imdb':1}))
imdb_list=[]
print(len(imdb))
for y in imdb :
    imdb_list.append(list(y.values())[0])

imdb_list[:10]

23541


[{'id': 6517, 'rating': 6.3, 'votes': 162},
 {'id': 8133, 'rating': 7.8, 'votes': 4680},
 {'id': 12512, 'rating': 6.8, 'votes': 489},
 {'id': 13025, 'rating': 7.8, 'votes': 3738},
 {'id': 15498, 'rating': 7.1, 'votes': 327},
 {'id': 11541, 'rating': 8.3, 'votes': 3942},
 {'id': 12349, 'rating': 8.4, 'votes': 56858},
 {'id': 14341, 'rating': 7.9, 'votes': 6646},
 {'id': 439, 'rating': 7.4, 'votes': 9847},
 {'id': 4150, 'rating': 5.8, 'votes': 223}]

# 1. Create New Collection

### Pipeline to Create New Collection
To create new collection named **clean_movies_hania**, we need to create pipeline that projects collection from movies_initial collection.<br>

In [106]:
pipeline = [
    {
        '$project' : {
            'title' : {"$cond": [{"$eq": ["$title", ""]}, "$nonExistinField", "$title"]},
            'year' : {"$cond": [{"$eq": ["$year", ""]}, "$nonExistinField", "$year"]},
            'num_mflix_comments' : '',
            'poster' : {"$cond": [{"$eq": ["$poster", ""]}, "$nonExistinField", "$poster"]},
            'plot' : {"$cond": [{"$eq": ["$plot", ""]}, "$nonExistinField", "$plot"]},
            'fullplot' : {"$cond": [{"$eq": ["$fullplot", ""]}, "$nonExistinField", "$fullplot"]},
            'rated' : {"$cond": [{"$eq": ["$rating", ""]}, "$nonExistinField", "$rating"]},
            'metacritic':{"$cond": [{"$eq": ["$metacritic", ""]}, "$nonExistinField", "$metacritic"]},
            'type' : {"$cond": [{"$eq": ["$type", ""]}, "$nonExistinField", "$type"]},
            'lastupdated' : {"$cond": [{"$eq": ["$lastupdated", ""]}, "$nonExistinField", "$lastupdated"]},
            
            'runtime' : {"$cond": 
                            [{'$eq': ["$runtime", ""]}, 0 ,{'$arrayElemAt':[{'$split': [ "$runtime", " min" ]}, 0]}]
                       },
            
            # array
            'directors' : {"$cond": [{"$eq": ["$director", ""]}, "$nonExistinField", {'$split': ["$director", ", "]}]},
            'cast' : {"$cond": [{"$eq": ["$cast", ""]}, "$nonExistinField", {'$split': ["$cast", ", "]}]},
            'writers' : {"$cond": [{"$eq": ["$writer", ""]}, "$nonExistinField", {'$split': ["$writer", ", "]}]},
            'genres' : {"$cond": [{"$eq": ["$genre", ""]}, "$nonExistinField", {'$split': ["$genre", ", "]}]},
            'languages' : {"$cond": [{"$eq": ["$language", ""]}, "$nonExistinField", {'$split': ["$language", ", "]}]},
            'countries' : {"$cond": [{"$eq": ["$country", ""]}, "$nonExistinField", {'$split': ["$country", ", "]}]},
            
            # convert released type from string to date
            'released': {
                '$cond': {
                    'if': {'$ne': ["$released", ""]},
                    'then': {
                        '$dateFromString': {
                            'dateString': "$released"
                        }
                    },
                    'else': ""}},
            
            # nested document
            'imdb' : {
                'id' : '$imdbID',
                'rating' : '$imdbRating',
                'votes' : '$imdbVotes',
                },
            
            'awards' : {
                'win' : {'$arrayElemAt' : [{'$split': [{'$arrayElemAt':[{'$split': ['$awards','win']}, -2]}, ' ']}, -1]},
                'nomination' : {'$arrayElemAt' : [{'$split': [{'$arrayElemAt':[{'$split': ['$awards','nomination']}, -2]}, ' ']}, -1]},
                'text' : '$awards',
                },
            
            'tomatoes': {
                'viewer': {'rating': "", 'numRewviews': "", 'meter': "" },
                'critic': {'rating': "", 'numRewviews': "", 'meter': "" },
                'dvd': "",
                'lastUpdated' : "",
                'production' : "",
                'rotten' : "",
                'fresh' : ""},
        }
    },
    {
        '$match':{'imdb': 
                      {'$in':[dict(sorted(imdb_list[x].items(), key=lambda x: x[0])) for x in range(len(imdb_list))]}
                 }
    },
    {
        '$out' : 'clean_movies_hania'
    }
]

clear_output ()
pprint.pprint(list(client_movies.sample_mflix.movies_initial.aggregate(pipeline)))

[]


In [23]:
# import pymongo
# from pymongo import MongoClient, UpdateOne
# import pprint
# from datetime import datetime
# import re
# from IPython.display import clear_output

# runtime_pat = re.compile(r'([0-9]+) min')
# updates = []
# count = 0

# for movie in client_movies.sample_mflix.clean_movies_hania.find({}):

#     fields_to_set = {}
#     fields_to_unset = {}

#     for k,v in movie.copy().items():
#         if v == "" or v == [""]:
#             del movie[k]
#             fields_to_unset[k] = ""

#     if 'director' in movie:
#         fields_to_unset['director'] = ""
#         fields_to_set['directors'] = movie['director'].split(", ")
#     if 'writer' in movie:
#         fields_to_unset['writer'] = ""
#         fields_to_set['writers'] = movie['writer'].split(", ")
#     if 'genre' in movie:
#         fields_to_unset['genre'] = ""
#         fields_to_set['genres'] = movie['genre'].split(", ")
#     if 'language' in movie:
#         fields_to_unset['language'] = ""
#         fields_to_set['languages'] = movie['language'].split(", ")
#     if 'country' in movie:
#         fields_to_unset['country'] = ""
#         fields_to_set['countries'] = movie['country'].split(", ")
        
#     if 'fullplot' in movie:
#         fields_to_unset['fullplot'] = ""
#         fields_to_set['fullPlot'] = movie['fullplot']
#     if 'rating' in movie:
#         fields_to_unset['rating'] = ""
#         fields_to_set['rated'] = movie['rating']

#     imdb = {}
#     if 'imdbID' in movie:
#         fields_to_unset['imdbID'] = ""
#         imdb['id'] = movie['imdbID']
#     if 'imdbRating' in movie:
#         fields_to_unset['imdbRating'] = ""
#         imdb['rating'] = movie['imdbRating']
#     if 'imdbVotes' in movie:
#         fields_to_unset['imdbVotes'] = ""
#         imdb['votes'] = movie['imdbVotes']
#     if imdb:
#         fields_to_set['imdb'] = imdb

#     if 'runtime' in movie:
#         m = runtime_pat.match(movie['runtime']) 
#         if m:
#             fields_to_set['runtime'] = int(m.group(1))

#     update_doc = {}
#     if fields_to_set:
#         update_doc['$set'] = fields_to_set
#     if fields_to_unset:
#         update_doc['$unset'] = fields_to_unset
    
#     updates.append(UpdateOne({'_id': movie['_id']}, update_one))
    
#     count += 1
#     pprint.pprint(update_doc)

# #     db2.clean_movies_hania.update_one({'_id': movie['_id']}, update_doc)

# 2. Validation Collection

### Connect to sample_mflix database to get clean_movies_hania collection

In [98]:
# show db sample_mflix
db2 = client_movies.sample_mflix #database
col_mov = db2.clean_movies_hania #movies collection

In [99]:
### show all db in cluster ###
print(client_movies.list_database_names())

### show all collection in used db ###
print(db2.list_collection_names())

['movies_list', 'sample_mflix', 'admin', 'local']
['clean_movies_devita', 'clean_movies_adipta', 'movies_initial', 'clean_movies_hania', 'clean_movies_firli', 'clean_movies_tio', 'clean_movies_isra', 'clean_movies_febi']


From the above, we can know that clean_movies_hania collection was succesfully created.

### A. Check whether all document in clean_movies_hania collection are same as movies collection
The flow to match the document :
1. Get document from movies collection sort by title ascending
2. Get document from clean_movies_hania collection sort by title ascending
3. Match clean_movies_hania's document to movies's document

In [100]:
val=col_movies.find().sort('title') #movies collection
val_mov=col_mov.find().sort('title') #clean_movies_hania collection

if val==val_mov :
    print("Document is matching.")
else :
    print("Document is not matching.")

Document is not matching.


From the above, we can know that **document in clean_movies_hania collection is ``not matching``** with movies collection.

### B. Check total documents of clean_movies_hania

In [101]:
# show all documents in clean_movies_hania
total = list(col_movies.find({}, {'_id':0, 'imdb':1})) #document in movies collection
total_mov = list(col_mov.find({}, {'_id':0, 'imdb':1})) #document in clean_movies_hania collection

print("Total documents in movies : ", len(total))
print("Total documents in clean_movies_hania : ", len(total_mov))

Total documents in movies :  23541
Total documents in clean_movies_hania :  23539


From the above, we can know that total number of **documents in clean_movies_hania is ``23539``**, fewer 2 documents than movies collection. 

In [102]:
# check why?
coll1 = list(col_movies.aggregate([{"$sortByCount":"$imdb"}])) #collection movies
coll2 = list(col_mov.aggregate([{"$sortByCount":"$imdb"}])) #collection clean_movies_hania

for i in coll1:
  if i not in coll2:
    print(i)

{'_id': {'id': 5, 'rating': 6.2, 'votes': 1189}, 'count': 2}


From the above, we can know that imdb with id 5 has 3 copies. It causes the number of documents in clean_movies_hania has 2 fewer documents.

### C. Check whether fields in clean_movies_hania can be found in movies collection
The flow to match the field :
1. Get unique fields from movies collection
2. Get unique fields from clean_movies_hania collection
3. Match clean_movies_hania's fields to movies's fields

In [103]:
# show movies collection field
fields=set([j for i in [list(doc.keys()) for doc in col_movies.find({})] for j in i])
fields

{'_id',
 'awards',
 'cast',
 'countries',
 'directors',
 'fullplot',
 'genres',
 'imdb',
 'languages',
 'lastupdated',
 'metacritic',
 'num_mflix_comments',
 'plot',
 'poster',
 'rated',
 'released',
 'runtime',
 'title',
 'tomatoes',
 'type',
 'writers',
 'year'}

In [107]:
# show clean_movies_hania collection field
field_mov=set([j for i in [list(doc.keys()) for doc in col_mov.find({})] for j in i])
field_mov

{'_id',
 'awards',
 'cast',
 'countries',
 'directors',
 'fullplot',
 'genres',
 'imdb',
 'languages',
 'lastupdated',
 'metacritic',
 'num_mflix_comments',
 'plot',
 'poster',
 'rated',
 'released',
 'runtime',
 'title',
 'tomatoes',
 'type',
 'writers',
 'year'}

In [108]:
# show movies collection
# field=(list(col_movies.find_one()))
# fieldmov=(list(col_mov.find_one()))

# check both field
if field_mov!=z :
    print("Fields are not matching.")
else :
    print("Fields are matching.")

Fields are matching.


From the above, we can know that **all fields in clean_movies_hania ``can be found``** in movies collection.

### D. Check whether all values in clean_movies_hania collection are same as in movies collection
Because of the number of document in both collection doesn't same, so I will use documents exclude imdb id 5 to compare.<br>
The flow to compare the value :
1. Get document from movies and clean_movies_hania collection exclude imdb id 5.
2. Compare clean_movies_hania's document with movies's document.

##### 1. Get document exclude imdb id 5

In [109]:
# total of movies document exclude imdb id 5
findmovies = col_movies.count_documents({ "imdb.id": {"$ne": 5 }})
findmovies

23538

In [110]:
# total of clean_movies_hania document exclude imdb id 5
findmov = col_mov.count_documents({ "imdb.id": {"$ne": 5 }})
findmov

23538

##### 2. Compare document from movies and clean_movies_hania collection

In [111]:
findmovies=col_movies.find().sort('title') #movies collection
findmov=col_mov.find().sort('title') #clean_movies_hania collection

if findmovies==findmov :
    print("The value of document is matching.")
else :
    print("The value of document is not matching.")

The value of document is not matching.


In [112]:
# values from movies
for mov in findmovies.limit(1):
    pprint.pprint(mov)

{'_id': ObjectId('573a13dcf29313caabdb2734'),
 'awards': {'nominations': 1, 'text': '1 nomination.', 'wins': 0},
 'cast': ['Rhys Wakefield', 'Logan Miller', 'Ashley Hinshaw', 'Natalie Hall'],
 'countries': ['USA'],
 'directors': ['Dennis Iliadis'],
 'fullplot': 'Three college friends hit the biggest party of the year, where a '
             'mysterious phenomenon disrupts the night, quickly descending '
             'into a chaos that challenges their friendships -- and whether '
             'they can stay alive.',
 'genres': ['Sci-Fi', 'Thriller'],
 'imdb': {'id': 2395385, 'rating': 5.5, 'votes': 5626},
 'languages': ['English'],
 'lastupdated': '2015-09-11 00:09:26.053000000',
 'metacritic': 60,
 'plot': 'Three college friends hit the biggest party of the year, where a '
         'mysterious phenomenon disrupts the night, quickly descending into a '
         'chaos that challenges their friendships -- and whether they can stay '
         'alive.',
 'poster': 'https://m.media-amazon.

In [113]:
# values from clean_movies_hania
for mov in findmov.limit(1):
    pprint.pprint(mov)

{'_id': ObjectId('5db005a9d0a60ca4acd15eb4'),
 'awards': {'nomination': '', 'text': '1 nomination.', 'win': None},
 'cast': ['Rhys Wakefield', 'Logan Miller', 'Ashley Hinshaw', 'Natalie Hall'],
 'countries': ['USA'],
 'directors': ['Dennis Iliadis'],
 'fullplot': 'Three college friends hit the biggest party of the year, where a '
             'mysterious phenomenon disrupts the night, quickly descending '
             'into a chaos that challenges their friendships -- and whether '
             'they can stay alive.',
 'genres': ['Sci-Fi', 'Thriller'],
 'imdb': {'id': 2395385, 'rating': 5.5, 'votes': 5626},
 'languages': ['English'],
 'lastupdated': '2015-09-11 00:09:26.053000000',
 'metacritic': 60,
 'num_mflix_comments': '',
 'plot': 'Three college friends hit the biggest party of the year, where a '
         'mysterious phenomenon disrupts the night, quickly descending into a '
         'chaos that challenges their friendships -- and whether they can stay '
         'alive.',
 'post

---

# Conclusion
In this practice case, it can conclude :
1. Projection was able to done (create clean_movies_hania from movies_initial collection).
2. **clean_movies_hania** collection has **``23539 documents``** (fewer than movies collection) because of duplication from imdb id 5 which has 3 copies in movies_initial collection.
3. All **fields** in clean_movies_hania collection are **``same``** as movies collection.
4. The **value** from clean_movies_hania collection **``will not always same``** as movies collection because of lack data from movies_initial such as :
    - no num_mflix_comments field
    - no wins, nomination, and text in awards field
    - no tomatoes field

---